core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -386,7 +387,7 @@ public final void runInContext() {
             }
           } catch (Throwable t) {
             stream.cancel(Status.CANCELLED.withCause(t).withDescription("Failed to read message."));
-            return;
+            Throwables.propagateIfPossible(t);
```

Comment:
> The result from the server was fine; it was the client processing code that was bolluxed up.

Yep. If the client is buggy, we reserve the right to fail the call even if the server responded correctly to notify the client it is buggy. Status.OK will _only_ come from a server, but other status codes may be generated locally.

> Re: the onError comment, are you thinking that an exception thrown in an onNext handler causes a call to onError?

Yes. Although I'm also saying I think that is broken if we've already received a status from the server. Which means it is very broken most of the time since unary calls are most frequent and unary calls have the status immediately after the message.

> So that starts to conflate/combine local exceptions with server/transport-generated errors? Hrm ...

Yeah. I'm expecting worst-case we'll add a field to Status that say whether it came from the server or was generated locally. Most code _should not_ care (if it is, it's probably wrong). However that bit of information is useful for a human debugging. Note that there are status codes that [we guarantee the library will not generate](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md). Having a specification for it is a bit new, so we've not yet added JavaDocs about the guarantee.

> And I still have concerns about any processing/delay in notification of Errors which can be pretty severe ...

Yeah, I can agree with that. Although our behavior is pretty similar to what happens with Futures and something like `ExecutorService.submit()` or `ScheduledExecutorService.schedule()`.

I've been trying to get an error handling system that is consistent, useful, and predictable. One of my main guiding principles was that an error should only be handled one place; we shouldn't log and then deliver an error expecting an application to handle it. If doing a streaming call and onNext() throws an exception, it seems the only sane thing is to cancel the call since the application will very possibly throw on later onNext() calls. If the call is cancelled, I feel like we really need to provide error details in the Status. But that also implies not logging since applications are intended to process the Status. So although I agree it's very disappointing that Errors may not be logged, I'm not sure where the weakness in the logic or assumptions is.

But I'm also aware that things get a little weird when considering server, because it basically behaves the opposite: exceptions cancel the call but also log.